### ISSUE TITLE
WEB3.0 BLOCKCHAIN BROWSER

### ISSUE THREAD

**Issue by jumezurike** [filecoin-project/devgrants#1194]
Open Grant Proposal: Cyberlockx - AFA browser completely on Filecoin network, IPFS gateway(filestore) and NFT storage

** Name of Project: WEB3.0 Blockchain Browser

** Proposal Category:** app-dev

** Proposer:**@jumezurike

** Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?** Yes

# Description of the project (WEB3.0 BLOCKCHAIN BROWSER)

We understand that Filecoin does not have any web 3/0 browser that could possibly enhance and indirectly enforce adoption with benefits for all (content) creators at this time: This is a conclusion we surmised from researching, talking to one of our clients as well as Filecoin engineers and developers. 
However, the process of making sure that the right data is stored in the right location by the right entity at the right time is something we find interesting. Any modification of our modern internet relies on this essence, of course the privacy and security is a factor we must not neglect.
The hashing proof is golden because it allows the right person to pull or download that data later on following the URI. Encrypting all data is something new and that is the objective of our proposition. We will like to introduce homomorphic and polymorphic encryption to information(data) on a grass root level, layer and beyond.
Filecoin's infrastructure fits very well with what want to build: A web 3/0 application program with an inherent security into the world wide web of today. We practically think that the time is ripe for encrypting everything online. This was not possible in the web 2.0 era; today it seems to be a way forward. 
If we must usher in a decentralized, distributed, secure and interoperable future web. We do not want to build a software because we want to mimic and mystify the cyber kill-chain. We want to do the very opposite by redefining and encrypting data from the onset with the power to compute over intermediate representation or ciphertext. 
Advanced authentication will be used to maintain origin identification.- Secure true digital identity a new dimension to physical to virtual representation. There will be a backup system to this; Data Access Security and Recovery solution (DASARS). We will build AFA an application program or a web 3/0 browser.- The mosaic for sensory context of the future internet.

**Problem:**
The internet was not designed initially with security in mind because the fundamental properties of the web protocols driving the internet today and are acceptable as safe, secure, and sound are NOT. As a result we have cyber criminals and the attacks that come with them. With the increased adoption of the hyper-information in mobile, IoT and DLT other emerging technologies we expect to hear more horror stories. Moreso, businesses are built around trading on consumers information. There is literarily no practical privacy preservation and information security built in in any browser in the world. Users do not have full control of their information and therefore cannot gain anything except through a third party.- The FAANGS business model shows that users cannot optimize any profit with their very own information

**Solution:**
A browser to use identity components for securing information by encrypted verification, validation, and evaluation. Users will have absolute control of their information. This will eventually demystify cyber security because security is everyone's problem. You break the solution into the following. 

- 1. Origin identification building cross-identity between physical and digital.
- 2. Ownership of information as control becomes autonomous( you share at your volition)
- 3. Search on encrypted data which curbs scraping and data mining.
- 4. Destruction of the cyber kill-chain which starts with reconnaissance.  
- 5. Removes third parties from trading other's information; data are encrypted.  

This will work perfectly well for Filecoin network which by design will become the most secure, decentralized record keeper for all sharable user data in all transactions. This will bring a dimension to both custodial and non-custodial wallets. There will be options for users to hold their communication exchange  object locally or externally for global reach. To top it all off, the browser cache hack will be gone since the caches are encrypted without key management of any kind. 

* On-demand     ----------------       Encryption is on demand
* Keyless       ----------------       No keys are moved, stored, reused, and managed.
* End-to end    ----------------       No link encryption and no third parties.
* Timeless security ------------       It is a quantum resistant cryptography variant.

### PREREQUISITE
* A. Core code; runtime engine (ECSMID)
* B. SDK
* C. API
* D. Libraries
* E. Filecoin network, IPFS Nodes/gateway, and NFT storage
* F. Extensions/plugins
* G. FVM and other Bots agents

All the required assets are pre-exiating images or pre-packaged as images to build componets of the AFA browser: We hope to use Filecoin network,IPFS nodes/gateway, and NFT storage as the proper decentralized file store for all clients with covert channels between nodes. **Some of the preexisting images are proprietary.** 

[Find the flow diagram from figma](https://www.figma.com/proto/VnUoj63xRyUcaFgeIYKBEc/AFA?page-id=0%3A1&node-id=1%3A587&viewport=230%2C650%2C0.47&scaling=contain&starting-point-node-id=1%3A587) 

![Find image of the AFA Browser](https://github.com/ubongndoh/CyberLock/blob/main/AFA.jpg) 

**AFA Browser**
[WALK THROUGH VIDEO ECSMID V 1.0.0](https://cyberlockx.xyz/docs/introduction)

### LokDon ECSMID V 1.0.0 USE CASES / FUNCTIONALITIES

[AWS ec2 free tier ECSMID V 1.0.0](https://aws.amazon.com/marketplace/pp/prodview-ksptmagbgdxqg)

[Docker ECSMID V 1.0.0](https://hub.docker.com/r/lkdn2022/lokdonecsmidv1sdk)

# Value
 * What are the benefits to getting this right?
   - This project will:
        - Improve both user and advnaced user (developer) experience and security for Filecoin nodes, network and drive more adoption.
        - Increase the value of Filecoin as the distributed storage with comparative browser needed to sharing expereince (data).
        - Generate many insights on how to improve performance, data control and business pace on Filecoin.
        - Drive improvement of state actors related to DID for origin identification which subsequently sets path for regulation before the policy shows up.
        - Bring new members (banks, fintech startup, cryptocurrency companies) and basicaly help in buidling new profiles (digital twins) for anyone with a smartphone in the globe to the Filecoin community 
        - In time of emergency and loss of life for the purpose of data access security and recovery solution (DASARS). 
        - This will remove user hardware requirements, providing easy backp and recovery ow wallet in emergency or loss of life.
        - This will drive users to do with their information what they see fit since they have absolute control absent third parties.
        -  Adding to Filecoin web 3.0 status. FIL wallet privacy and security will be protected with quantum resistant cryptography: On-demand key less end-to-end encryption. More transaction will be recorded on Filecoin because the independent ad free, private, sercure, non-tracking and cooky free AFA browser. 
        -  Filecoin will become the most secure Web 3/0 network with the algorithm underlying the FVM. No doubt it will well equiped for control and sales of data to brokers who will now send them to the ingestors like google, FB and etc. 
        -  We will provide the first search engine that works on encrpted data; making biographic and demographic information scarce enough to cause users to pay on-demand once their search becomes available. THe encrypted information (output) will have to be paid for the user to view or copy it: This implies the originator of the data will have to fix a price or batter it. 
        -  Users are notified when searches are done and by who. They will be notified of the search and that will be open since the data is encrypted. The option to approve the transaction sales of stop it. This will quickly increase adoption. 
        -  The first web 3/0 browser to make password legacy, with an enhanced password system and assistive technology for those in the far end of the spectrum. 
* What are the risks if you don't get it right?
  - Obervations include:
      - Expensive retrieval deals may scare away some developers.
      - Clients or providers may waste FIL if data transfers aren't properly executed.
      - A failed attempt at developing a successful integrate of the browser into the network may decrease the community's confidence in Filecoin becoming a major storage solution for web 3.0.
      - Failure to heed guidance of well architected framework may lead to dead ends, wasting of time which could decimmate funds to finish the project.
* What are the risks that will make executing on this project difficult?
  - Basically trust levels:
      - This is a complex technical project with many moving parts, permisions and access control which we hope we can solve with better architecture. 
      - The proposed architecture is new, but we need to proceed gradually, methodically and iterate in order to advance our insight.
      - There will be need to spend more time on ipfs with hope that it will update and sync data from many entities in a record time.
      - The os and rendering engine are not easy but we can tweak chromium to achieve our goal. 

# Deliverables

Requirement is clear: 
* A. secure Software development lifecycle will be used all through the process stages.
- i. Requirement    -------------------------- Risk assessment
- ii. design        -------------------------- Threat model design review
- iii. development  -------------------------- Static analysis
- iv. Testing       -------------------------- Security testing and code review
- v. Deployment     -------------------------- Security assessment and secure configuration
- vi. CI/CD         -------------------------- CI/CD
**We move straight to design.**
* 1. The user interface and user experience design: this includes the address bar, back/forward button, bookmarking menu, etc. Every part of the browser display except the window where you see the requested page. Threat model and requirement review.
* 2. The browser engine: marshals actions between the UI and the rendering engine.

* 3. The rendering engine: responsible for displaying requested content. For example if the requested content is HTML, the rendering engine parses HTML and CSS, and displays the parsed content on the screen.
* 4. Networking: for network calls such as HTTP requests, using different implementations for different platform behind a platform-independent interface. -Internal Testing.

* 5. UI backend: used for drawing basic widgets like combo boxes and windows. This backend exposes a generic interface that is not platform specific. Underneath it uses operating system user interface methods.
* 6. JavaScript interpreter. Used to parse and execute JavaScript code.
* 7. Data storage. This is a persistence layer. The browser may need to save all sorts of data locally, such as cookies. Browsers also support storage mechanisms such as localStorage, IndexedDB, WebSQL and FileSystem.- Filecoin storage will be used. - More advance developer testing, validation and security tuning. 

* 8. Filecoin integration, extension, bot-agents, FVM integration with ECSMID for pipeline/network security by encryption. 
* 9. Add utilities like secure true digital id, payment wallets, video conferencing, Filecoin individual cloud and NFT minting. - Public testing, we plan several iteration s before this testing prior deployment to production. Testing will continure in the sandbox. 
* 10. Competitive price structure at about 0.5% for any payments on the browser. Subscription will be charge for the utilities once you go over the free limit. 
* 11. Deploy perpetual Beta to production 

# Development Roadmap

## January 1 - March 23 2023 (Q1)
Requirement is clear: We will use Secure SDLC
A. secure Software development lifecycle (used all through the process stages)
- i. Requirement    -------------------------- Risk assessment
- ii. design        -------------------------- Threat model design review
- iii. development  -------------------------- Static analysis
- iv. Testing       -------------------------- Security testing and code review
- v. Deployment     -------------------------- Security assessment and secure configuration
- vi. CI/CD         -------------------------- CI/CD
* We move straight to design.
* 1. The user interface and user experience design: this includes the address bar, back/forward button, bookmarking menu, etc. Every part of the browser display except the window where you see the requested page. Threat model and requirement review.
* 2. The browser engine: marshals actions between the UI and the rendering engine.

## March 24 - June 4 2023 (Q2)
* 3. The rendering engine: responsible for displaying requested content. For example if the requested content is HTML, the rendering engine parses HTML and CSS, and displays the parsed content on the screen.
* 4. Networking: for network calls such as HTTP requests, using different implementations for different platform behind a platform-independent interface. -Internal Testing.

## June - September 28 2023 (Q3)
* 5. UI backend: used for drawing basic widgets like combo boxes and windows. This backend exposes a generic interface that is not platform specific. Underneath it uses operating system user interface methods.
* 6. JavaScript interpreter. Used to parse and execute JavaScript code.
* 7. Data storage. This is a persistence layer. The browser may need to save all sorts of data locally, such as cookies. Browsers also support storage mechanisms such as localStorage, IndexedDB, WebSQL and FileSystem.- Filecoin storage will be used. - More advance developer testing, validation and security tuning. 
*Advance authentication that encourages web 3 users to engage e.g Enhance password systems https://ccnft.lokdon.com
a special kind of NFT will be used to validate asset, gain access and secure resources.
 
* ## September 29 - December 28 2023 (Q4)
* 8. Filecoin integration, extension, bot-agents, FVM integration with ECSMID for pipeline/network security by encryption. 
* 9. Add utilities like secure true digital id, payment wallets, video conferencing, Filecoin individual cloud and NFT minting. - Public testing, we plan several iteration s before this testing prior deployment to production. Testing will continure in the sandbox. 
* 10. Competitive price structure at about 0.5% for any payments on the browser. Subscription will be charge for the utilities once you go over the free limit. 
* 11. Deploy perpetual Beta to production 

# Total budget requested

  * Budget will cover 10 engineers and 1 Security solution architect living expenses for the duration of the project
        12 months of sprint development - $890K
  * Infrastructure for running multiple Filecoin nodes:
        12 months of billings - $50K
  * Total: $940K

# Maintenance and Upgrade Plans

   The browser software program or application will be powered by Filecoin and ECSMID. We will continue to maintain and improve them as the browser evolves moving forward.

# Team
Team Members

   - Joisah Umezurike - Application Security Engineer / Solution Architect
   - Nahom Wosenu Bogale - Principal Engineer
   - Alexander Ibekason - Backend Engineer
   - Utkarsh Jaiswal - Frontend developer / UI-UX designer
   - Faraz Naliwar - Mobile Application developer (Flutter dart/android)
   - Alend Jarjis - Project manager/QA
   - Ubong Ndoh - React native developer
   - 7 more sofware engineers with experience in browser OS, search engine and rendering engine.
   - 3 software engineers with background native web languages security, browser security and Filecoin VM. 

# Team Member LinkedIn Profiles

*  https://www.linkedin.com/in/josiah-umezurike/
*  https://www.linkedin.com/in/nahom-wosenu-80a234160/
*  https://www.linkedin.com/in/alexander-ibekason/
*  https://www.linkedin.com/in/utkarshjaiswal1997/
*  https://www.linkedin.com/in/faraznilawar/
*  https://www.linkedin.com/in/alend-jarjis-95b6b6144/
*  https://www.linkedin.com/in/ubongndoh/

# Team Website

    https://cyberlockx.xyz

# Relevant Experience

We have over 40 years experience as a team. Strong background in cyber security and domain expertise.
We have worked together for at least 2 years on an average. More full time employees will be added for this project. 
We know the vision and what we need done.  I and my co-founder have been working together for 5 years.

# Team code repositories
+ https://github.com/ubongndoh/CyberLock.git 

# Additional Information

You can find our research paper on [Researgate](https://www.researchgate.net/publication/336242914_Lattice-Face_Key_Infrastructure_LFKI_for_Quantum_Resistant_Computing)

--The metaverse is a place where we create a world in our own image.
This requires a revamp of the web and its protocols. 
Web 3/0 is an attempt into a secure and private internet.
It promises humanity read, write and ownership of data 
However, it is not without its challenges.
It certainly has seed phrases flying around. 
More so, keys flying all over the place. 
What is securing those seed phrases?
It ought to be some other keys.
What secures those keys?
The answer, infinite loop. 
This seems like a very big challenge.
Keyless approach could save the day.--

**Comment by jumezurike** [comment]
@realChainLife I wanted to know if there is something wrong with our grant application. We have not heard from you. Let us know if there is something amiss. 
Thanks.

**Comment by ErinOCon** [comment]
Hi @jumezurike, thank you for reaching out! This proposal is still under review. We will be in touch with an update or questions once our initial review is complete.

**Comment by jumezurike** [comment]
We have finalized this https://ccnft.lokdon.com  on Goerli network (testnet) files on filecoin.  It could serve as a part of the browsers optional password enhancement system.
This kind of color coded/tagged NFT are unique carrying the payload of the origin following advanced authentication. It can help a lot of people in the far end of the spectrum as well as digital content creators.

**Comment by jumezurike** [comment]
we have completed the authentication piece. Take an look and enjoy it. 
https://ccnft.lokdon.com
I hope you get back to us soon we are ready to go.

**Comment by ErinOCon** [comment]
Hi @jumezurike, I know it has been a considerable amount of time since we have provided an update for this proposal. In light of the macroeconomic climate, the review of our budget and priorities for the fiscal year resulted in a longer evaluation period - thank you for all of your patience!

After much consideration, we have determined that we will not be moving forward with a grant at this time. We are wishing you all the best as you continue to build!
